### DISCUSSION TITLE
Issue with chunk lighting when entering new instance

### DISCUSSION THREAD

**Discussion by GreyLake** [Minestom/Minestom#805] [Help] [answered]
Hey! So sorry to bother, I'm still getting into Minestom and am having an issue when sending players to various instances. I'm aware that Minestom currently doesn't have a lighting issue, But I think think that's what's causing the issue. 

Let's say I create two instanceContainers, such as "lobby" and "example", the players default spawning instance is set to "lobby";

`
         Map<String, WorldInstance> instances = new HashMap<>();
         String instanceName;
         InstanceManager manager = new InstanceManager();

        {
            instanceName = "lobby";
            // Create the instance
            instances.put(instanceName, manager.createInstanceContainer(DimensionDefault.getDimensionType()));
            instances.get(instanceName).setChunkLoader(new LoaderDefault("instances/lobby"));
            instances.get(instanceName).setChunkGenerator(new GeneratorFlat());
            instances.get(instanceName).enableAutoChunkLoad(true);

            instances.get(instanceName).saveInstance();
            instances.get(instanceName).saveChunksToStorage();
        }

        {
            instanceName = "example";
            // Create the instance
            instances.put(instanceName, manager.createInstanceContainer(DimensionDefault.getDimensionType()));
            instances.get(instanceName).setChunkLoader(new LoaderDefault("instances/example"));
            instances.get(instanceName).setChunkGenerator(new GeneratorFlat());
            instances.get(instanceName).enableAutoChunkLoad(true);

            instances.get(instanceName).saveInstance();
            instances.get(instanceName).saveChunksToStorage();
        }
`

Alright cool, so now when the server starts, it creates two nice simple instances with a flat terrain (The terrain generator is default). LoaderDefault() is an exact replica of AnvilLoader(), it was implemented into a class for code review and future modifications. DimensionDefault.getDimensionType() is just a replica of OVERWORLD, but with the ability to change ambient light, etc.
Now when a players joins the server, they spawn at the center (0, 44, 0) of "lobby". Here there is no lighting issues, however, now let's say I want to send the player to another instance, so we create a command '/teleport <instance name>'.

`
var teleportArgument = ArgumentType.String(teleportWorldCommandTypeMessage);

        addSyntax((sender, context) -> {
            Player player = (Player)sender;
            String teleportInstance;

            if(context.has(teleportArgument)){
                teleportInstance = context.get(teleportArgument);
                if(Ghost.getInstance(teleportInstance) != null){
                    InstanceContainer instance = Ghost.getInstance(teleportInstance);
                    player.setInstance(instance);
                    player.teleport(new Pos(0, 44, 0));
                }
            }
        }, teleportArgument);
`

I know, it's not the best way to teleport a player, but it works. This in theory should teleport the player to the center of any instance we specify it to (assuming the instance exists and has been created). Now when the player teleports to this region, this is where we encounter the issue. In a radius of about 20 chunks (varies with every try) there is no SkyLight at all. It is recorded as 0. BlockLight does work, and once we leave the 'unlighted' region, we have SkyLight. Now, I'm still new to this, and I'm sorry if I'm just being an idiot and literally missed something simple. I've tried loading in the chunk at center before the player joins by using `instance.loadChunk(new Pos(0, 44, 0);` however this also yielded the same results. I've tried implementing other peoples "world generating" implementations, but non of them seemed to change the outcome. 

Now I know there's a "quick fix" which is to set the ambiantLight to 2.0f, but that doesn't fix future issues such as mobspawning, and underground lighting (away from SkyLight). It's a fix, but it's a client side fix, and not a server fix (It may cause issues and even more trouble later on when I try working on other things). I could add a light engine, not too sure how (or where I'd inject it), but light seems to behave properly out of the effected chunks around spawn. 

I hope someone can help me out, I'd really appreciate it. You can definitely consider me a "noob" and still in the process of learning. I'm willing to try a lot of code just to get this to work, or to discuss possible fixes/new implementations. Anyway, thanks in advance! 

ADDED:
Also! When teleporting back to the "lobby" instance, the issue appears there as well. Another note: The issue appears to occur even further than just spawn, when teleporting to a new world, if you adventure out you'll find large patches of unlit chunks. Starting to really think this might be because of a lack of light engine, but then again - Why does it work find when the player joins the first "lobby" world for the first time, there's no lighting issue then.

TEMPORARY SOLUTION COMMENTED! 
View my comment for a Jank temporary fix. This should only be used if you want the effect of proper lighting while the lighting engine is being worked on.

**Comment by TheMode** [comment, marked as answer]
As you said, this is caused due to Minestom not having a light engine. And we plan to add one in time (more contributors wanted!)

Here is a branch I made: https://github.com/Minestom/Minestom/tree/light
And a port of starlight that we unfortunately cannot use due to licensing: https://github.com/Minestom/Minestom/pull/771

**Reply by GreyLake** [reply]
Cool! thank you so much for the reply, I could try for now using the port of starlight while I wait for the light branch to develop (I don't think the light branch at the moment works to any significant amount, more specifically the `createLightPacket()`, then again though, I have no idea how to implement any light engine into an instance. I'm gonna try messing around with the ported 'ish-stable' (good enough just to work on things). If you do have any info on how I could implement the ported starlight version, definitely let me know! As of right now, I taking a while guess and saying it's already implemented because InstanceContainer has it's own LightManager objects and uses it when loading chunks.

Wish I could contribute, but I'm not experienced enough - As you can tell! Maybe one day.

**Reply by GreyLake** [reply]
Well, that isn't working - ha

**Comment by GreyLake** [comment]
If anyone wants a quick fix for this, simply create a Chunk Supplier and Chunk matching the same code as a Dynamic Chunk (Unless you want to implement your own Chunk). In the methos `createLightData()`, simply replace it for:
`
 private LightData createLightData() {
        BitSet skyMask = new BitSet();
        BitSet blockMask = new BitSet();
        BitSet emptySkyMask = new BitSet();
        BitSet emptyBlockMask = new BitSet();
        List<byte[]> skyLights = new ArrayList<>();
        List<byte[]> blockLights = new ArrayList<>();

        int index = 0;
        for (Section section : sections) {
            index++;
            final byte[] skyLight = new byte[2048];
            System.out.println("Filled Light");
            Arrays.fill(skyLight, (byte) 255);
            final byte[] blockLight = section.getBlockLight();
            //if (skyLight.length != 0) {
                skyLights.add(skyLight);
                skyMask.set(index);
            //} else {
            //    emptySkyMask.set(index);
            //}
            if (blockLight.length != 0) {
                blockLights.add(blockLight);
                blockMask.set(index);
            } else {
                emptyBlockMask.set(index);
            }
        }
        return new LightData(true,
                skyMask, blockMask,
                emptySkyMask, emptyBlockMask,
                skyLights, blockLights);
    }
`
It's not the best fix, and should ONLY be used temporarily until a lighting engine is in place. This temporarily fixes lighting issues when placing/breaking blocks as well as changing between instances. If you're too lazy, you can use my already edited code and add it to your project through Jitpack.io -> SOTAMC/SOTAChunks : Modified chunk class is SOTAChunkSupplier(). For loading a world after saving, you can perform the same modification (already performed for SOTAChunkLoader()).
